kernel/src/transaction.rs

```diff
@@ -0,0 +1,508 @@
+    /// commit and visible to anyone who describes the table history.
+    pub fn operation(&mut self, operation: String) {
+        self.operation = Some(operation);
+    }
+
+    /// WARNING: This is an unstable API and will likely change in the future.
+    ///
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. The engine data passed here must have exactly one row, and we
+    /// only read one column: `engineCommitInfo` which must be a map<string, string> encoding the
+    /// metadata.
+    ///
+    /// The engine is required to provide commit info before committing the transaction. If the
+    /// engine would like to omit engine-specific commit info, it can do so by passing pass a
+    /// commit_info engine data chunk with one row and one column of:
+    /// 1. `engineCommitInfo` column with an empty Map<string, string>
+    /// 2. `engineCommitInfo` null column of type Map<string, string>
+    ///
+    /// Any other columns in the data chunk are ignored.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>) {
+        self.commit_info = Some(commit_info.into());
+    }
+}
+
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+// TODO(zach): in order to make the returning of a transcation useful, we need to add APIs to
+// update the transaction to a new version etc.
+#[derive(Debug)]
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, Version),
+}
+
+// given the engine's commit info we want to create commitInfo action to commit (and append more actions to)
+fn generate_commit_info(
+    engine: &dyn Engine,
+    operation: Option<&str>,
+    engine_commit_info: Arc<dyn EngineData>,
+) -> DeltaResult<Box<dyn EngineData>> {
+    if engine_commit_info.length() != 1 {
+        return Err(Error::InvalidCommitInfo(format!(
+            "Engine commit info should have exactly one row, found {}",
+            engine_commit_info.length()
+        )));
+    }
+
+    let timestamp: i64 = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("time went backwards..?")
+        .as_millis()
+        .try_into()
+        .map_err(|_| Error::generic("milliseconds since unix_epoch exceeded i64 size"))?;
+    let commit_info_exprs = [
+        // TODO(zach): we should probably take a timestamp closer to actual commit time?
+        Expression::literal(timestamp),
+        Expression::literal(operation.unwrap_or(UNKNOWN_OPERATION)),
+        Expression::literal(Scalar::Struct(StructData::try_new(
```

Comment:
why does this not serialize as:
```json
{
   "operationParameters": {
     "operationParameters": null
  }
}
```

Reply:
because unfortunately when we have `"something": null` it won't write it at all (similar to how we have top-level action columns `"remove": null` etc.)

Reply:
I've isolated the hack to `generate_commit_info` and we can use the 'real' schema for `CommitInfo` now!

Reply:
Got it. Thanks